---
title: "Transcriptomics: Lecture 2"
subtitle: |
  | Biotech 7005/Bioinf 3000
  | Frontiers of Biotechnology: Bioinformatics and Systems Modelling
  | The University of Adelaide
author: "Dr Stevie Pederson (They/Them)<br>stevie.pederson@thekids.org.au"
email: "stevie.pederson@thekids.org.au"
institute: |
  | Black Ochre Data Labs
  | The Kids Research Institute Australia
date: "2025-09-15"
date-format: "long"
bibliography: transcriptomics/references.bib
format: 
  revealjs:
    theme: [transcriptomics/custom.scss]
    height: 720
    width: 1280
    sansfont: Times New Roman
    slide-number: c
    show-slide-number: all
    include-after-body: transcriptomics/uoa_bodl_logos.html
  html: 
    css: [transcriptomics/custom.scss]
    output-file: transcriptomics1-lecture2.html
    embed-resources: true    
    toc: true
    toc-depth: 2
    include-in-header: transcriptomics/uoa_bodl_logos.html
---

## Helpful Links

- The slides are available [here](./transcriptomics-lecture2-slides.html)
- The same notes are available as a single-page html [here](./transcriptomics1-lecture2.html)


## 

::: {.text-red}
<br>
*I'd like to acknowledge the Kaurna people as the traditional owners and custodians of the land we know today as the Adelaide Plains, where I live & work.*
<br><br>
*I also acknowledge the deep feelings of attachment and relationship of the Kaurna people to their place.*
<br><br>
*I pay my respects to the cultural authority of Aboriginal and Torres Strait Islander peoples from other areas of Australia, and pay our respects to Elders past, present and emerging, and acknowledge any Aboriginal Australians who may be with us today*

:::

# RNA-Seq

## RNA Sequencing

According to @Wang2009-hf

> RNA-Seq, also called RNA sequencing, is a particular technology-based sequencing technique which uses next-generation sequencing (NGS) to
reveal the **presence** and **quantity** of RNA in a biological sample at a **given moment**, analyzing the continuously changing cellular transcriptome.


## RNA Sequencing {.slide-only .unlisted}

:::: {.columns}

::: {.column}
- Microarrays are still published regularly
    + Also used extensively for methylation
- RNA sequencing is now the dominant technology

::: {.fragment}

- Strong improvement for:
    + transcript-level resolution
    + un-annotated genes 
    + genomic variants
    + allelic bias
    
:::
    
:::
::: {.column}


```{r}
#| echo: false
#| message: false
#| fig-width: 6
#| fig-height: 6
library(tidyverse)
theme_set(theme_bw())
here::here("Lectures/transcriptomics") %>% 
  fs::dir_ls(glob = "*search.csv") %>% 
  lapply(read_csv) %>% 
  bind_rows(.id = "platform") %>% 
  mutate(
    platform = platform %>% 
      basename() %>% 
      str_remove_all("_search.csv") %>% 
      str_to_title() %>% 
      str_replace_all("Rna", "RNA-")
  ) %>% 
  ggplot(aes(Category, `Publications (total)`)) +
  geom_line(aes(colour = platform), linewidth = 1) +
  ggrepel::geom_label_repel(
    aes(label = label, colour = platform),
    data = . %>% 
      dplyr::filter(Category == max(Category), .by = platform) %>% 
      mutate(
        label = paste0(
          Category, "\n", scales::comma(`Publications (total)`))
      ),
    force_pull = 0, nudge_y = -1000,
    show.legend = FALSE, size = 4, fill = "#FFFFFF90"
  ) +
  ggtitle("Publication search by app.dimension.ai on 26-08-2025") +
  labs(x = "Year", colour = "Platform") +
  scale_y_continuous(labels = scales::comma) +
  scale_x_continuous(
    breaks = seq(2000, as.integer(format(Sys.Date(), format = "%Y")) - 1, by = 5)
  ) +
  scale_colour_brewer(palette = "Set1") +
  theme(
    text = element_text(size = 16),
    plot.title = element_text(face = "italic", size = 14),
    legend.position = "inside",
    legend.position.inside = c(0.99, 0.01),
    legend.justification.inside = c(1, 0)
  )
```
:::
::::

## RNA Sequencing {.slide-only .unlisted}

- Microarrays rely on probes for transcripts defined *at design time*
- Restricted in the number of transcripts/genes targeted
    +  Gencode 48 (GRCh38): 78,686 genes + 385,669 transcripts
- Probes capture non-specific binding
    + Affymetrix use 25-mer probes
    + Illumina arrays use 60-mer probes $\implies$ sequences better targeted
    
::: {.fragment}
[These limitations do not exist for RNA-Seq]{.text-red}
:::

## RNA Sequencing {.slide-only .unlisted}

- Directly sequence the biological material
- Map to most recent reference at any point in time
- Assemble a transcriptome (tissue specific)
- Detect InDels / SNPs in expressed sequences

::: {.fragment} 

- Multiple variations 
    + total-RNA or polyA transcripts [$\implies$ most similar to microarrays]{.fragment}
    + small-RNA libraries
    + Long Reads (Oxford Nanopore, PacBio) [<br>$\implies$ originally isoform discovery, quantitative methods improving]{.fragment}
    
:::
    
## The RNA Population Of a Eukaryotic Cell

:::: {.columns}

::: {.column width='60%'}

![Image taken from @ijms19051310](transcriptomics/ijms-19-01310-g001.png){width='700px' fig-align="left"}

:::
::: {.column width='40%'}
<br>

- rRNA $\approx$ 80%[^1]
- tRNA $\approx$ 15% 
- All other RNA $\approx$ 5%

[^1]: https://bionumbers.hms.harvard.edu/bionumber.aspx?s=n&v=5&id=100264
:::
::::


## The Key Steps

- Focus from here on will be sequencing mRNA using short reads

:::: {.columns}
::: {.column}

::: {.fragment}
1. Library Preparation
    + RNA Quality assessment <br>(i.e. RNA degradation)
    + Selecting target molecules
    + Adding sequencing primers
:::

::: {.fragment}
2. Sequencing 
3. Alignment + Quantitation
:::
:::
::: {.column}

::: {.fragment}
4. DE Gene Detection
5. Downstream Analysis
6. (Optional) Nobel Prize

:::
:::
::::

<!-- ## Library Preparation -->

<!-- - RNA is a single-stranded molecule $\implies$ easily degraded -->
<!--     + Degraded RNA can skew expression levels during comparisons -->
<!-- - RNA Integrity Number (RIN) commonly used to identify low-quality samples -->
<!--     + Assesses 28S:18S rRNA ratio, along with overall size distribution -->

## RNA Selection

1. Select for poly-adenylated RNA using oligo-dT-based methods
    + Only extracts intact mRNA with a polyA tail (includes some ncRNA)

:::: {.columns}

![Image from https://www.lexogen.com/polya-rna-selection-kit/](transcriptomics/03polya_workflow1.jpg){width='950px' fig-align="left"}

::::

## RNA Selection  {.slide-only .unlisted}

2. Enzymatically deplete rRNA sequences 
    + rRNA targeted using probes $\implies$ dsRNA degraded
    + Can additionally target hbRNA (whole blood)

![Image from https://support.illumina.com.cn](transcriptomics/Ribo_Zero_1.jpg){fig-align="left"}

    


## Library Preparation


- RNA is then fragmented and size selected (200-300nt)
    + Very short transcripts always lost during this step
- cDNA produced
- Sequencing adapters added
    + Indexes are unique to each individual library $\implies$ always have replicates
    + Optionally contain Unique Molecular Identifiers (UMI)<br>$\implies$ Helps identify PCR duplicates
- Most RNA-Seq now retains *strand-of-origin* information (Stranded RNA-Seq)
    + During PCR only the first cDNA template retained

    
## Library Preparation {.slide-only .unlisted}

![Image courtesy of https://www.lubio.ch/blog/ngs-adapters](transcriptomics/NGS-adapters_01.png)

## Library Preparation {.slide-only .unlisted}

- Each double-stranded cDNA then has the final structure (+UMI)

![Image courtesy of https://www.lubio.ch/blog/ngs-adapters](transcriptomics/csm_NGS-Y-adapters_a9077b39d0.jpg)

## Sequencing

:::: {.columns}

::: {.column width='60%'}

![Image taken from https://www.bmkgene.com/dnarna-sequencing-illumina-sequencer-product/](transcriptomics/A8.jpg){fig-align="left"}
:::
::: {.column width='40%'}
:::
::::



# Alignment and Quantitation

## Genomic Alignment

- Alignment to a reference genome requires a *splice-aware* aligner
    + A GTF (Gene Transfer File) required when building the index<br>$\implies$ Provides all exon-transcript-gene co-ordinates
    + New Gencode, Ensembl etc releases at regular intervals
- Most common aligners are STAR [@Dobin2013-us] & hisat2 [@Kim2019-ib]
    + Return alignments as a `bam` file
- Aligned reads are then counted to provide gene-level counts
    + htseq [@htseq2014] and featureCounts [@Liao2014-gy] are very common
    + The same GTF should be used as during indexing
    
    
## Genomic Alignment  {.slide-only .unlisted}

![Image from https://ucdavis-bioinformatics-training.github.io/2019_March_UCSF_mRNAseq_Workshop/data_reduction/alignment.html](transcriptomics/index_igv17.png)

## Counting Alignments

:::: {.columns}

::: {.column width='50%'}
![Image taken from https://htseq.readthedocs.io/en/latest/htseqcount.html](transcriptomics/count_modes.png){width='85%' fig-align='left'}

:::
::: {.column width='50%'}
- Some alignments align beautifully within exon structures
- Some overhang a little
    + Unspliced mRNA?
- Some genes are overlapping
    + Stranded libraries can resolve
- Maybe bacterial reads span genes within an operon

:::

::::

## Gene-Level Counts

- The region encoding a gene is (relatively) well defined
- An alignment within a gene is easy to assign to that gene

::: {.fragment}
- Much more difficult to identify which transcript it came from
    + Many transcripts share multiple exons
- Splice Junctions were the earliest approach

:::

## Gene-Level Counts {.slide-only .unlisted}

![Image taken from @Yu2015-f5](transcriptomics/gkv608fig1.jpeg)

## Transcriptome Alignment

- An alternative is to provide a *reference transcriptome*
    + Alignments no longer need to be splice aware
- Reads can (& commonly do) align to multiple transcripts
- Much faster than traditional alignment

::: {.fragment}

- Pseudo-alignment is used by kallisto [@Bray2016-qq] 
- Statistically modelled expression estimates used by salmon [@Patro2017-sn]
- Return *transcript-level counts* without bam files
- Add transcript-level counts $\implies$ gene-level counts

:::

## Pseudo-Counts

- Salmon counts are actually pseudo counts output by model fitting
    + Predicts the proportion of library derived from transcript
    + Fitted using EM-algorithm or Bayesian modelling

::: {.incremental}
- Counts bootstrapped to provide uncertainty estimates of prediction<br>$\implies$ *Measures how confident we are in the transcript-level counts*
- Transcript-level counts can be scaled by uncertainty estimate [@Baldoni2024-zf] when performing DTE analysis
:::

# Differential Gene Expression Analysis

## Differential Gene Expression Analysis

- A fundamental question:

**Does the overall abundance of a gene differ between experimental conditions?**

::: {.fragment}
- Need a statistical approach to answer this question
:::


## Count-Based Data

- Under both reference-types $\rightarrow$ counts represent expression
- These are *discrete* data (i.e. not continuous values)
- Microarrays were continuous values (fluorescence intensity)
    + Modelled using log~2~-transformed values $\implies \mathcal{N}(\mu, \sigma)$
    + Linear regression, $t$-tests etc
    + Mean and variance are independent variables
    
::: {.fragment}
- Count data is commonly modelled using a Poisson Distribution $\implies \text{Poisson}(\lambda)$
    + Poisson variance is defined as being equal to the mean i.e. $\sigma^2 = \mu$<br>$\implies$Mean and variance *are not independent variables*
    
:::

## Count-Based Data {.slide-only .unlisted}

- In most cases the observed variance is greater than the mean$\implies$ *over-dispersion*

::: {.fragment}
- We use the Negative Binomial distribution to model counts for gene $g$ in sample $i$

$$
y_{gi} \sim \text{NB}\left(\mu_{gi}, \sigma^2_g(\mu_{gi} +\phi\mu^2_{gi})\right), \text{where } \phi > 0
$$
:::
::: {.fragment}

- Can be thought of as like a Poisson with extra variation
- Extra variation is strictly defined in quadratic relationship to mean
    + Sometimes described as a combination of *technical* & *biological* variation

:::

## Count-Based Data {.slide-only .unlisted}

- Fit *generalised linear models* (GLMs) to model counts and estimate logFC
- Implemented in edgeR [@Chen2016-ig] and DESeq2 [@love2014-ds]
    + Slight differences in handling of over-dispersions
    + Both default to FDR-adjusted $p$-values
    
::: {.fragment}
- Same model for transcripts, but dividing counts by uncertainty estimate [@Baldoni2024-zf]
+ Can also model transcripts as proportions of gene-level estimates
:::

## Count-Based Data {.slide-only .unlisted}

- Poisson/NB-GLMs fit the rate of an event, i.e. counts per fixed measurement window
- Sequencing data produces 'libraries' of counts $\rightarrow$ total counts = *library size*
- In model fitting $\rightarrow$ estimate rate as a function of library size
    
::: {.fragment}
- TMM (edgeR) or RLE (DESeq2) approaches estimate scaling factors
    + Moderates the effect of highly expressed genes which dominate library size
    + In a perfect world $\implies$ all scaling factors = 1
- Library size when modelling is multiplied by scaling factors
    + The most effective normalisation methods for RNA-Seq
    
:::

## Count-Based Data {.slide-only .unlisted}

- A common alternative measure is *counts per million* (CPM) [or logCPM]{.fragment}
- CPM is simply counts divided by (library size / 1,000,000)
    + No longer discrete $\implies$ continuous data
- Relationship between mean and variance *still retained* for logCPM
    + Can't use naively in classic `limma`-based linear models

::: {.fragment}
- The limma-voom method [@Law2014-xq] uses weights to break the mean-variance relationship
    + Can then assume normally-distributed data >$\implies$ slightly more flexible models
- Alternatively, the `limma-trend` works effectively
:::

<!-- Put a volcano & MA plot here, maybe with tables -->

## Enrichment Testing

:::: {.columns}


::: {.column}
![](transcriptomics/apoptosisDAG.png)
:::


::: {.column}
- Often want to identify key pathways associated with results
    + Are genes from specific pathways or gene-sets most impacted

::: {.fragment}
- Gene Ontology (GO) Terms:
    + Carefully constructed, hierarchical database of terms
    + Biological Process, Molecular Function and Cellular Component
    
:::
:::

::::


## Enrichment Testing {.slide-only .unlisted}

:::: {.columns}


::: {.column width="60%"}
![Oxidative Phosphorylation pathway as annotated in the KEGG database](transcriptomics/hsa00190.png){width="95%"}
:::


::: {.column width='40%'}
- Kyoto Encyclopedia of Genes and Genomes (KEGG):
    + Molecular pathways with complete topology
    
:::

::::

## Enrichment Testing {.slide-only .unlisted}

:::: {.columns}
::: {.column width="60%'}

- Often use approaches based on Fisher's Exact Test
    + Compare enrichment of DE genes in a pathway vs not DE Genes in pathway
    + Not-DE $\equiv$ Background
- Pathways often share DE genes $\implies$ biological expertise
- Multiple Testing across 1000s of pathways



:::
::: {.column width='40%'}
::: {.text-red}
|       | DE | Not DE | %DE |
|:----- | --------:| ------------:| ----:|
| In Pathway | 10 | 40 | 20% |
| Not In Pathway | 990 | 14000 | 6.7% |

::: {.fragment}
- `r scales::percent(40/14040, 0.01)` not DE genes in pathway
- `r scales::percent(10/1000, 1)` of DE genes in pathway
:::
:::
:::
::::

## Enrichment Testing {.slide-only .unlisted}

::: {style='font-size:95%'}

:::: {.columns}
::: {.column width='45%'}
- An alternative approach is to use ranked lists
    + Gene Set Enrichment Analysis (GSEA)
- No requirement to classify genes as DE or not DE

::: {.fragment}
- Walk along the ranked list of genes increasing score if gene in gene-set
    + In the example, begin at the right
    + Largest deviation from zero is -ve

:::
:::
::: {.column width='55%'}
![Image taken from clusterProfiler vignette https://yulab-smu.github.io/clusterProfiler-book/chapter12.html](transcriptomics/gseaplot2-1.png)
:::
::::

:::



# Beyond Differential Expression

## Transcript Assembly

::: {.notes}
StringTie is the quick & dirty method. Will also turn up some weird artefacts
:::

:::: {.columns}
::: {.column width='60%'}

![Image taken from [@Pertea2015-zd]](transcriptomics/stringtie.png){fig-align="left"}
:::
::: {.column width='40%'}

- With a good reference genome $\implies$ *StringTie* can identify novel transcripts
- Un-annotated genes/lncRNA
:::
::::




## Transcript Assembly {.slide-only .unlisted}

- StringTie returns a GTF with novel transcripts added to reference (See [here](https://ccb.jhu.edu/software/stringtie/index.shtml?t=manual#output))
    + Gene & Transcript annotations capture the entire transcribed region
    + Exons are annotated by transcript + gene
- Also returns counts for DE analysis


## Transcript Assembly {.slide-only .unlisted}

- Complete *de novo* transcriptome assembly using Trinity [@Grabherr2011-bh]
- Can alternatively perform a full *reference-guided assembly* 
    + Also works with more distantly-related references
    <!-- + Built a sea-snake venom-gland transcriptome using another snake genome -->

::: {.incremental}
- Three basic modules
    1. Inchworm: Naively assembles reads into contigs
    2. Chrysalis: Pools contigs into *de Bruijn* graph
    3. Butterfly: Trims *de Bruijn* graph and compares against raw reads
- Check assembly against known conserved orthologs (BUSCO)
<!-- - Predict protein function from sequence homology -->

:::

## Transcript Assembly {.slide-only .unlisted}

:::: {.columns}
::: {.column width='60%'}

![*An example de Bruijn graph* (Wikimedia commons)](transcriptomics/debruijn.jpg)
:::
::: {.column width='40%'}
- Reads represent fragmented transcripts
- Reads broken into shorter *k*-mers
- Adjacent *k*-mers $\rightarrow$ graph
- Wise choice of *k* helps extend graph well beyond read length
- Well-represented paths through graph $\rightarrow$ transcripts
:::
::::

## Transcript Assembly {.slide-only .unlisted}

![Figure from @Haas2013-ma](transcriptomics/trinity.jpg){fig-align="left"}

## Long Read Technology

- Most transcriptome assemblies performed using Trinity/StringTie
- Long Reads are now becoming the dominant platform 
    + Oxford Nanopore (ONT) reads from 50 bp to >4 Mb
    + Pacific Biosciences (PacBio) up to 25kb
    + Illumina maxes out around 2x150nt
- Both can sequence complete transcripts!
    
## Long Read Technology {.slide-only .unlisted}

- Much more error prone
    + Originally ~15% but now approaching Illumina error-rates
    + Same molecule repeatedly sequenced
- Repeat sequencing makes quantitation more challenging<br>$\implies$ Now becoming viable

<!-- ## WGCNA -->

<!-- ## WGCNA {.slide-only .unlisted} -->

# High Resolution Technologies

## Single-Cell Transcriptomics

- Single-Cell RNA-Seq is becoming a dominant transcriptomics platform
    + Conventional RNA-Seq is now sometimes called *bulk* RNA-Seq
- Enables insights into highly heterogeneous samples (e.g. immune cells)
- Originally polyA-selected 3' sequences $\rightarrow$ full-length
- Incomplete transcript capture $\implies$ many missing genes in each cell

## Single-Cell Transcriptomics {.slide-only .unlisted}

![Figure taken from @Haque2017-ae](transcriptomics/scRNAWorkflow.png)

## Single-Cell Transcriptomics {.slide-only .unlisted}

- Single-Cell expression patterns used to cluster similar cells
    + Clustering can be highly subjective
    + Key marker genes identified to identify cell-types
- Often visualised using PCA, tSNE or UMAP to summarise expression values
- Clusters annotated as key cell-types
    + Can be *pseudo-bulked* for DE comparison between cell-types within sample
    
::: {.fragment}
- Methods such as CITE-Seq also label cells using tagged antibodies
    + Cell types identified from cell-surface markers
    
:::

## Spatial Transcriptomics 

- Spatial Transcriptomics uses tissue slices on a microscope slide
- Each spot on the slide contains oligo-dT probes with a unique spatial barcode
    + Spatial barcode helps assign reads to a tissue region
- Interaction between key cell-types can be identified

::: {.fragment}
- Some techniques use Fluorescence In Situ Hybridisation (FISH) to detect transcripts
    + Cell boundaries detected in combination with classic stains
    + Transcript location *within the cell*
- Can also cluster cell-types like scRNA then map back to tissue
- Multiple tissue layers possible $\rightarrow$ \$\$\$
:::



## Spatial Transcriptomics {.slide-only .unlisted}

![Figure from @Zhang2023-sa](transcriptomics/41586_2023_6808_Fig4_HTML.png)


# References

##
